docs/RFCS/20211207_graceful_draining.md

```diff
@@ -0,0 +1,422 @@
+`connection_wait` until we confirm that customers have to apply this feature.
+
+In summary, the desired final state of this design is to have the following
+timeline for the draining:
+- `drain_wait` (unchanged): `health?ready=1` starts to return the draining
+  state of the node, and load balancer listening to this http for health check
+  stops routing new SQL connections to this draining node. But new SQL
+  connections from other channels are still allowed, and existing SQL
+  connections still work.
+- `connection_wait`: No new SQL connections are allowed. Existing SQL
+  connections still work, but those without queries in-flight will be given a
+  timeout to be recycled by connection pool or closed by user. Idle SQL
+  connections whose lifetime is longer than the connection pool’s maximum
+  lifetime will be recycled.
+  SQL connections not returned to the pool won’t be recycled. If all SQL
+  connections are closed, it does an early exit before the timeout, and proceed
+  to `lease_transfer_wait`.
+- `query_wait` (unchanged): Any SQL connections without open queries in-flight
+  will be closed. Any SQL connection, once its query is finished, will be
+  closed. If all SQL connections are closed, the server does an early exit
+  before the timeout, and proceed to `lease_transfer_wait`. Otherwise, the
+  server waits till `query_wait` times out, and then force into the
+  `lease_transfer_wait` phase.
+  In this case, all on-going queries will be left uncommitted.
+- `lease_transfer_wait`: the server waits to transfer range leases.
+
+# Demo with changes from Technical Design
+[A 4-min demo recording][improvement recording] with the same Hikari connection
+pool setting can be found here, where we did manual testing.
+
+To add an automated test, we will refer to `pkg/sql/conn_executor_test.go`
+and `pkg/sql/pgwire/pgwire_test.go` to set up a connection pool, and apply
+a timeline similar to the one in the demo.
+
+Note that in the demo we only manually tested with a single-node server. A
+multi-node server with a load balancer involved should also be considered.
+The `drain_wait` should still be set longer than the time that the load
+balance needs to stop routing new SQL connections for this draining node, and
+users just need to set the length of `connection_wait` longer than the
+maximum lifetime of the connection pool.
+
+Specifically, `drain_wait` should be longer than the `check inter` in the
+[http health check of HAProxy][HAProxy health check],
+and the `periodSeconds` in [k8s’s readinessProbe setting][k8s health check].
+
+# Possible Stretch
+We may also want to enable the server to only allow new SQL connections of
+certain types during `drain_wait`, since some customers mentioned that they’d
+like to reserve access to the node when the draining process started.
+The possible solution is to add a gate in the connection register that
+looks at a connection’s metadata (e.g. if it’s from admin or not) during
+`drain_wait`.
```

Comment:
this could even change to allow admins to login during connection_wait also, but that is also just a possible future extension